Add support for spatial point values to the batch importer

henriknyman opened the pull request:
- Add simple map-parser for PointValue.
- Move the logic for creating a PointValue with different default
coordinate reference systems depending on the given fields, so that
it can be shared with the batch importer without having to instantiate
an actual Map object. Instead it can build an array with the given
fields directly from the parser.
- Let the csv module depend on the values module so it can produce
values directly.

henriknyman committed:
Add support for spatial point values to the batch importer

- Add simple map-parser for PointValue.
- Move the logic for creating a PointValue with different default
coordinate reference systems depending on the given fields, so that
it can be shared with the batch importer without having to instantiate
an actual Map object. Instead it can build an array with the given
fields directly from the parser.
- Let the csv module depend on the values module so it can produce
values directly.

henriknyman committed:
Fix test result verification for properties

henriknyman committed:
Fail on duplicate point value fields

henriknyman committed:
Fix code style

sherfert commented on community/csv/src/main/java/org/neo4j/csv/reader/Extractors.java:
```diff
@@ -85,6 +90,7 @@
     private final Extractor<long[]> longArray;
     private final Extractor<float[]> floatArray;
     private final Extractor<double[]> doubleArray;
+    private final PointExtractor point_;
```
Is there a particular reason this field ends with an underscore?

henriknyman replied:
No, I guess I based it on a copy of one that had an underscore :)

MishaDemianenko replied:
oh :no_good: looks like we missing some checkstyle rule :nut_and_bolt:

sherfert commented on community/kernel/src/main/java/org/neo4j/kernel/impl/util/ValueUtils.java:
```diff
@@ -338,4 +274,18 @@ public static RelationshipValue fromRelationshipProxy( Relationship relationship
+     */
+    public static Value asValue( Object value )
+    {
+        if ( value instanceof Value )
```
`Values.of` checks if the argument is a Point, before checking if the argument is a Value. Hence, passing `PointValue`s works. Is this method really needed?

One reason I can imagine is performace, since `Values.of` will create copies of passed `PointValue`s. If that is the reason for this method, it should be documented as such.

henriknyman replied:
Oh, you are right that `Values.of` happens to work for PointValue (and temporal values). But it seems to go against the original purpose of the method according to its documentation. Also we don't want to create new objects unnecessarily during batch import. 

We may probably want to move all extractors over to producing `Value`s directly (now that we have the module dependency), in which case the conversion through this method will not be needed anymore.

fickludd replied:
+1 on creating `Value`s directly instead of going via java Boxed objects.

sherfert commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputBatchImportIT.java:
```diff
@@ -452,6 +475,34 @@ private void buildUpExpectedData(
+            // Special verifier for point property
+            Consumer verifyPoint = actualValue ->
+            {
+                // The y-coordinate should match the node number
```
Is the test simply set up like this or is there another reason why Y is equal to the node number?

henriknyman replied:
It was just set up like that to get some controlled variation of the lines

sherfert commented on community/values/src/main/java/org/neo4j/values/storable/PointValue.java:
```diff
@@ -255,4 +261,268 @@ public boolean withinRange( PointValue lower, boolean includeLower, PointValue u
+
+    public static PointValue fromMap( MapValue map )
+    {
+        AnyValue[] fields = new Value[PointValueField.__MAX_VALUE__.ordinal()];
```
It feels like you could as well use `PointValueField.values().length` and get rid of this `__MAX_VALUE__` altogether.

sherfert commented on community/values/src/main/java/org/neo4j/values/storable/PointValue.java:
```diff
@@ -255,4 +261,268 @@ public boolean withinRange( PointValue lower, boolean includeLower, PointValue u
+        X
+                {
+                    @Override
+                    ValueGroup valueType()
```
This becomes way shorter and more readable when you create a field in this enum and a Constructor, instead of overriding the same method everywhere.

E.g.:

```
 private enum PointValueField
    {
        X( NUMBER );

        private ValueGroup valType;

        PointValueField( ValueGroup valType )
        {
            this.valType = valType;
        }

        ValueGroup valueType()
        {
            return ValueGroup.NO_VALUE;
        }
    }
```

sherfert reviewed (changes requested):
I have some minor questions and comments. All in all it looks good to me!

Please have a look at the test failures in `SpatialFunctionsAcceptanceTest`

henriknyman committed:
Fix failing tests

- Adapt to new error messages
- LazyMap.get should return null instead of throw exception on
non-existing property key

henriknyman committed:
Fix review comments

sherfert commented:
Your latest changes seem to have introduced other test failures. Apart from that it looks good to me now!

henriknyman committed:
Fix failing tests
